File: lecture09/04_height/height.py
#!/usr/bin/env python3

import collections

# Constants

INFINITY = float('inf')

# Structures

Node = collections.namedtuple('Node', 'value left right')

# Functions

def height_tree(root):
    ''' Use divide and conquer to compute the height of binary tree '''
    # Base case: Non-existent node
    if not root:
        return 0

    # Divide and conquer: recursively solve left and right sub-trees
    left_height  = height_tree(root.left)
    right_height = height_tree(root.right)

    # Combine: take max sub-tree height and add 1
    return max(left_height, right_height) + 1

# Main Execution

if __name__ == '__main__':
    # Create tree
    root = Node(7,
        Node(5,
                Node(3, Node(6, None, None), None),
                None),
        Node(4,
                None,
                Node(2, None, None),
    ))

    # Compute height of tree
    print(height_tree (root))

File: lecture09/04_height/height.cpp
// height_tree.cpp

#include <iostream>
#include <memory>

using namespace std;

// Node structures

template <typename T>
struct Node {
    T value;
    struct Node *left;
    struct Node *right;
};

// Find height of tree

template <typename T>
size_t height_tree(Node<T> *root) {
    if (root == nullptr) {
    	return 0;
    }

    size_t left  = height_tree(root->left);
    size_t right = height_tree(root->right);

    return max(left, right) + 1;
}

// Main execution

int main(int argc, char *argv[]) {
    Node<int> *root = new Node<int>{7,
    	new Node<int>{5,
    	    new Node<int>{3,
		new Node<int>{2, nullptr, nullptr},
		nullptr},
    	    nullptr
	},
    	new Node<int>{4,
    	    nullptr,
    	    new Node<int>{2, nullptr, nullptr}
	}
    };

    cout << height_tree(root) << endl;
    return 0;
}
